### PR TITLE
[SPARK] Extracts job name creation to providers

### PR DESCRIPTION
**Pull request by arturowczarek**
* Implements a job name provider for "park.openlineage.appName" property
* Implements a job name provider for "spark.app.name" property
* Implements a job name provider for AWS Glue

### Problem

AWS Glue is generating a new "spark.app.name" property with each execution. This causes the OpenLineage events consumers to generate a new diagram with each execution.

### Solution

This pull request detects if the `spark.app.name` was autogenerated by Glue and uses the Glue job name in such case.
The code was becoming difficult, so I extracted each job name provisioning strategy to a separate provider.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [x] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [x] Your comment includes a one-liner for the changelog about the specific purpose of the change (_not required for changes to tests, docs, or CI config_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [x] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2024 contributors to the OpenLineage project

### CODE REVIEW

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/AwsGlueJobNameProvider.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
+/*
+/* Copyright 2018-2024 contributors to the OpenLineage project
+/* SPDX-License-Identifier: Apache-2.0
+*/
+
+package io.openlineage.spark.agent.lifecycle;
```
Comment (arturowczarek) [inline integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/AwsGlueJobNameProvider.java:6]: Any suggestion where to put the providers? The package `lifecycle` seems a little bit off.

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/SparkApplicationExecutionContext.java [hunk 2]
```diff
@@ -144,14 +152,21 @@ private OpenLineage.ParentRunFacet buildApplicationParentFacet() {
...
-    return openLineage
-        .newJobBuilder()
-        .namespace(eventEmitter.getJobNamespace())
-        .name(normalizeName(name));
+    return openLineage.newJobBuilder().namespace(eventEmitter.getJobNamespace()).name(getJobName());
+  }
+
+  private String getJobName() {
+    return jobNameProviders.stream()
+        .map(provider -> provider.getJobName(olContext))
```
Comment (pawel-big-lebowski) [inline integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/SparkApplicationExecutionContext.java:160]: wouldn't it make sense to add extra method to an interface verifying if a job provider is applicable. Something like `isDefinedAt(OpenLineageContext context)`. Such an approach is common within this codebase.

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/AwsGlueJobNameProvider.java (hunk 1 above)
Comment (pawel-big-lebowski) [reply integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/AwsGlueJobNameProvider.java:6]: `spark.agent.job.providers` ?

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/SparkApplicationExecutionContext.java (hunk 2 above)
Comment (arturowczarek) [reply integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/SparkApplicationExecutionContext.java]: I was thinking about it. It requires checking the same parameters while checking if the provider is applicable and when it provides the value. But it should never be a performance bottleneck so I refactored the code.

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/SparkApplicationExecutionContext.java [hunk 3]
```diff
@@ -39,6 +45,11 @@ class SparkApplicationExecutionContext implements ExecutionContext {
   private final EventEmitter eventEmitter;
   private final OpenLineageRunEventBuilder runEventBuilder;
   private final OpenLineage openLineage = new OpenLineage(Versions.OPEN_LINEAGE_PRODUCER_URI);
+  private final List<JobNameProvider> jobNameProviders =
```
Comment (mobuchowski) [inline integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/SparkApplicationExecutionContext.java:48]: I think if we're hard forcing the precedence order, we need to comment it here.

**Review by mobuchowski — COMMENTED** [review]
Overall I love the idea. Have some nits about clarity/documentation though.

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/job/naming/JobNameProvider.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * There can be multiple ways how the job name can be obtained for OpenLineage initial event. It
+ * depends on the environment the application is deployed in, the properties passed to the app, etc.
+ * This interface provides the contract for each provider to either provide some job name or tell
+ * (with empty Optional) that it is unable to provide one.
+ */
+public interface JobNameProvider {
+
+  boolean isDefinedAt(OpenLineageContext openLineageContext);
```
[the remark this answers is not in the document]
Comment (arturowczarek) [reply integration/spark/app/src/main/java/io/openlineage/spark/agent/job/naming/JobNameProvider.java]: I don't see such a need. We are reading the properties from `SparkContext` or `SparkOpenLineageConfig` at the moment.

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/job/naming/JobNameProvider.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+package io.openlineage.spark.agent.job.naming;
+
+import io.openlineage.spark.api.OpenLineageContext;
+
+/**
+ * There can be multiple ways how the job name can be obtained for OpenLineage initial event. It
+ * depends on the environment the application is deployed in, the properties passed to the app, etc.
+ * This interface provides the contract for each provider to either provide some job name or tell
+ * (with empty Optional) that it is unable to provide one.
+ */
```
[the remark this answers is not in the document]
Comment (arturowczarek) [reply integration/spark/app/src/main/java/io/openlineage/spark/agent/job/naming/JobNameProvider.java]: I renamed it. Is it better now?

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/SparkApplicationExecutionContext.java (hunk 3 above)
Comment (arturowczarek) [reply integration/spark/app/src/main/java/io/openlineage/spark/agent/lifecycle/SparkApplicationExecutionContext.java]: I've moved this part to a separate class - `ApplicationJobNameResolver` where we have hardcoding part explained and the logic.

**Review by pawel-big-lebowski — APPROVED** [review]
Code looks good to me. 
Please resolve other discussions before merging.

**Review by mobuchowski — APPROVED** [review]

**Pull request merged**